buffer/src/main/java/io/netty/buffer/CompositeByteBuf.java

```diff
@@ -1712,30 +1723,26 @@ public CompositeByteBuf discardReadBytes() {
             return this;
         }
 
-        // Remove read components.
-        int firstComponentId = toComponentIndex0(readerIndex);
-        for (int i = 0; i < firstComponentId; i ++) {
-            Component c = components[i];
-            c.free();
-            if (lastAccessed == c) {
-                lastAccessed = null;
+        int firstComponentId = 0;
+        Component c = null;
+        for (int size = componentCount;
+                firstComponentId < size; firstComponentId++) {
+            c = components[firstComponentId];
+            if (c.endOffset > readerIndex) {
+                break;
             }
+            c.free();
         }
 
-        // Remove or replace the first readable component with a new slice.
-        Component c = components[firstComponentId];
-        if (readerIndex == c.endOffset) {
-            // new slice would be empty, so remove instead
-            c.free();
-            if (lastAccessed == c) {
-                lastAccessed = null;
-            }
-            firstComponentId++;
-        } else {
-            c.offset = 0;
-            c.endOffset -= readerIndex;
-            c.adjustment += readerIndex;
-            c.slice = null;
+        // Replace the first readable component with a new slice.
+        c.offset = 0;
+        c.endOffset -= readerIndex;
+        c.adjustment += readerIndex;
+        c.slice = null;
+
+        Component la = lastAccessed;
+        if (la != null && la.endOffset < readerIndex) {
+            lastAccessed = null;
         }
 
         removeCompRange(0, firstComponentId);
```

Comment:
@njhill could we make a separate PR for the `CompositeByteBuf` changes and so keep things focused ?

Reply:
Sure, the reason I included them here is that the other changes kind of rely on discarding read components to be cheap, especially if there are none to discard (as would be common). I'll split into another PR though which can precede this one.